### PR TITLE
Partial affinity

### PR DESCRIPTION
**Pull request by kriskowal**

### CODE REVIEW

File: service-proxy.js [hunk 1]
```diff
@@ -320,11 +327,73 @@ ServiceDispatchHandler.prototype.refreshServicePeer =
...
+    peer.connectTo();
+};
+
+ServiceDispatchHandler.prototype.refreshServicePeerPartially =
+function refreshServicePeerPartially(serviceName, hostPort) {
+    var self = this;
 
+    // Reset the expiration time for this service peer
     var time = self.channel.timers.now();
     self.exitServices[serviceName] = time;
```
Comment (kriskowal) [inline service-proxy.js]: restore in non-enabled case

File: service-proxy.js [hunk 2]
```diff
@@ -320,11 +327,73 @@ ServiceDispatchHandler.prototype.refreshServicePeer =
 function refreshServicePeer(serviceName, hostPort) {
     var self = this;
 
+    if (self.partialAffinityEnabled) {
+        return self.refreshServicePeerPartially(serviceName, hostPort);
+    }
+
+    // The old way: fully connect every egress to all affine peers.
     var peer = self.getServicePeer(serviceName, hostPort);
-    peer.connect(true);
```
Comment (kriskowal) [inline service-proxy.js]: separate into other diff

File: service-proxy.js [hunk 3]
```diff
@@ -320,11 +327,73 @@ ServiceDispatchHandler.prototype.refreshServicePeer =
...
+    // Obtain and sort the affine worker and relay lists.
+    var relays = Object.keys(self.egressNodes.exitsFor(serviceName));
+    relays.sort();
+    var serviceChannel = self.getOrCreateServiceChannel(serviceName);
+    var workers = serviceChannel.peers.keys();
+    workers.sort();
+
+    // Find our position within the affine relay set so we can project that
+    // position into the affine worker set.
+    var relayIndex = search(relays, self.channel.hostPort);
```
Comment (kriskowal) [inline service-proxy.js]: or binaryIndexOf

**Comment by kriskowal** [comment]
r @Raynos

File: test/partial-affinity.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    seedConfig: {
+        'hyperbahn.partialAffinity': {
+            minPeersPerRelay: 1,
+            minPeersPerWorker: 1
+        }
+    }
+}, function t(cluster, assert) {
+    runScenario(cluster, assert, onCompletion);
+    function onCompletion(err) {
+        assert.end(err);
```
Comment (Raynos) [inline test/partial-affinity.js]: in all cases please assert(numberOfExpectedSockets, N)

File: service-proxy.js [hunk 5]
```diff
@@ -77,6 +80,10 @@ function ServiceDispatchHandler(options) {
     });
     self.rateLimiterEnabled = options.rateLimiterEnabled;
 
+    self.partialAffinityEnabled = options.partialAffinityEnabled;
+    self.minPeersPerWorker = options.minPeersPerWorker || DEFAULT_MIN_PEERS_PER_WORKER;
+    self.minPeersPerRelay = options.minPeersPerRelay || DEFAULT_MIN_PEERS_PER_RELAY;
+
     self.egressNodes.on('membershipChanged', onMembershipChanged);
```
Comment (Raynos) [inline service-proxy.js]: Can refresh on membershipChanged ?

File: test/partial-affinity.js (hunk 4 above)
Comment (kriskowal) [inline test/partial-affinity.js]: The trouble is that we are not yet reaping unneeded peers, so we’ll expect any number of connections between partially connected and fully connected. The only case we can assert is invalid is “not connected”, for which there is ample evidence if the requests fail.

File: service-proxy.js (hunk 5 above)
Comment (kriskowal) [inline service-proxy.js]: Debatable whether that is useful. The affinity gets updated upon every advertise and these happen with some regularity. Doing the refresh on membership changes would only make the shift in traffic more sudden.

Comment (Raynos) [inline service-proxy.js]: #legit

File: service-proxy.js [hunk 6]
```diff
@@ -321,11 +328,100 @@ ServiceDispatchHandler.prototype.refreshServicePeer =
...
+            workers.length,
+            Math.max(
+                self.minPeersPerRelay,
+                self.minPeersPerWorker * ratio
+            )
+        )
+    );
+    var stop = (start + length) % workers.length;
+
+    self.logger.info('Refreshing service peer affinity', self.extendLogInfo({
```
Comment (Raynos) [inline service-proxy.js]: This log is O(n) advertise. 

Once we've proven the feature with confidence we don't want this.

We may want a log for "new peers" or "removing peers" though.

Hyperbahn should be silent under steady state. We can keep this here for now.

**Comment by jcorbin** [comment]
rebase and then let's merge this for a 2.11cut

**Comment by Raynos** [comment]
lgtm.

**Pull request merged**